### ISSUE TITLE
SFT 训练的方式需要加prompt吗？

### ISSUE THREAD

**Issue by sunnnnnnnny** [FunAudioLLM/CosyVoice#1514]
Hello, I noticed that in section 2.7 of the paper, the MSFT fine-tuning method adds a prompt "Speaker A<|endofprompt|>" before the text. This is different from the SFT fine-tuning method in the current examples. I would like to ask which of these two methods is better? I look forward to your reply.

**Comment by haziyevv** [comment]
I have trained the model like this. It learns the characteristics of each speaker, but it makes a lot of simple mistakes. It is worse than training without that prompt in my experience.

**Comment by sunnnnnnnny** [comment]
> I have trained the model like this. It learns the characteristics of each speaker, but it makes a lot of simple mistakes. It is worse than training without that prompt in my experience.

+1

**Comment by ScottishFold007** [comment]
Hey @sunnnnnnnny and @haziyevv,

As someone who’s been tinkering with CosyVoice for a while, I can share some thoughts on this prompt thing based on my own experiments. The `"Speaker A<|endofprompt|>"` trick from the CosyVoice 2 paper (Section 2.7) is meant to help the model nail down specific speaker vibes, especially when you’re dealing with multiple speakers or trying to keep the voice consistent across languages. It’s like giving the model a clear heads-up: “Yo, this is Speaker A, do your thing.” In my experience, this works great for zero-shot or cross-lingual tasks where you need the model to lock onto a specific speaker’s tone or style without mixing them up.

That said, I totally get @haziyevv’s point about it causing simple mistakes. I’ve seen this too—sometimes the model gets a bit too hung up on the prompt format and starts tripping over basic stuff, like mispronouncing words or sounding unnatural. It’s like it’s trying so hard to follow the “Speaker A” cue that it forgets how to just talk normally. When I trained without the prompt (like in the standard SFT examples, e.g., `examples/libritts/cosyvoice/run.sh`), the output felt smoother and less error-prone, especially for single-speaker setups or when the dataset already has clear speaker labels.

Here’s my take:
- **With the prompt**: It shines when you’re working with multiple speakers or need precise control over voice characteristics. I’ve had good results with it for cross-lingual stuff, like making a Chinese speaker sound natural in English. But you gotta make sure your dataset is solid, with varied examples, or the model might overfit to that exact prompt structure and mess up if you don’t use it during inference.
- **Without the prompt**: Feels more robust for simpler tasks or when you’re confident the speaker embeddings can handle the job. It’s less fussy and seems to generalize better if your input format varies.

For what it’s worth, I’ve found that mixing things up during training—like including prompts for some samples but not others—helps the model stay flexible. Also, tweaking the training script to lower the weight of the prompt’s loss (like what’s hinted at in some of the CosyVoice 3 docs) can cut down on those annoying errors @haziyevv mentioned. If you’re seeing a lot of mistakes, maybe check if your dataset has enough variety or if the prompt is throwing the model off during inference.

Honestly, it depends on what you’re aiming for. If you’re building something like a multi-speaker voice assistant, stick with the prompt to keep things tight. For a single-speaker setup, you might not need it. Try running a quick test with a small dataset both ways and check the WER or NMOS scores to see what’s working better for you. Let me know how it goes or if you want me to share some of my training configs!

**Comment by haziyevv** [comment]
Hey @ScottishFold007 , thanks for the detailed explanation — it was very helpful.

Here’s what I’ve tried so far:
	1.	Converted data to this format: Speaker_A prompt<|endofprompt|>text
	2.	Without speaker tags, but provided speaker info via utt2spk during input data creation.
	3.	Without any speaker info, assigning all audio to the same speaker (single-speaker setting).

Interestingly, the third experiment gave me the best results, while the first performed the worst. I always set config.use_spk_embedding: False — maybe that’s part of the reason. Other than that, I haven’t changed anything; I’m just using the default configs from libritts/cosyvoice2/run.sh.

If you could share the configs you used for your experiments, that would be really helpful for comparison.

**Comment by ScottishFold007** [comment]
> Hey [@ScottishFold007](https://github.com/ScottishFold007) , thanks for the detailed explanation — it was very helpful.
> 
> Here’s what I’ve tried so far: 1. Converted data to this format: Speaker_A prompt<|endofprompt|>text 2. Without speaker tags, but provided speaker info via utt2spk during input data creation. 3. Without any speaker info, assigning all audio to the same speaker (single-speaker setting).
> 
> Interestingly, the third experiment gave me the best results, while the first performed the worst. I always set config.use_spk_embedding: False — maybe that’s part of the reason. Other than that, I haven’t changed anything; I’m just using the default configs from libritts/cosyvoice2/run.sh.
> 
> If you could share the configs you used for your experiments, that would be really helpful for comparison.

Hey @haziyevv,

Thanks for the update and for sharing your experiments—it’s awesome to see you digging into CosyVoice like this! Your results are super insightful, and I’m glad to hear the single-speaker setup worked best for you. The fact that you’re using `use_spk_embedding: False` and getting better results without prompts makes sense, especially with what we know about CosyVoice’s design across versions. Since you mentioned some of your `config.yaml` settings (like `prompt_loss_weight` and `add_prompt_prob`) are custom tweaks not in the official repo, I’ll keep that in mind and tailor my suggestions to work with both your setup and CosyVoice’s core mechanics. Let’s dive into your experiments and how speaker embeddings (or lack thereof) play into this, and I’ll share some configs I’ve used that might help you fine-tune further.

### Unpacking Your Experiments with CosyVoice’s Design
Based on the latest info on CosyVoice (1.0, 2.0, and 3.0), the role of speaker embeddings has evolved, which sheds light on why your setups performed the way they did. Here’s a quick breakdown of your experiments in the context of CosyVoice’s architecture:

1. **Experiment 1: Speaker Prompt (`Speaker_A<|endofprompt|>text`)**  
   You reported this performed the worst, likely because you set `use_spk_embedding: False`. In CosyVoice 1.0, speaker embeddings (x-vectors extracted from Mel spectrograms) were used in both the language model (LM) and flow-matching model to capture timbre and style. However, in CosyVoice 2.0 and 3.0, the LM no longer uses speaker embeddings to avoid mixing language or paralinguistic info, which can mess with prosody naturalness (see CosyVoice 2 paper, arXiv:2412.10117). With embeddings disabled, the model relies solely on the prompt to differentiate speakers, but without enough training data variety or a robust prompt strategy, it can overfit to the prompt format, leading to those “simple mistakes” you and @haziyevv noticed. 
   
2. **Experiment 2: No Prompt, Speaker Info via `utt2spk`**  
   This setup should theoretically work well, as `utt2spk` provides speaker metadata that CosyVoice’s flow-matching model can use to generate speaker-specific Mel spectrograms. But with `use_spk_embedding: False`, the model isn’t leveraging these embeddings in the flow-matching stage either, which might explain why it didn’t outperform the single-speaker setup. The flow-matching model in CosyVoice 2.0 and 3.0 relies heavily on speaker embeddings to map semantic tokens to timbre-specific audio, so disabling them likely reduced speaker differentiation, making results similar to a prompt-less setup.

3. **Experiment 3: Single-Speaker, No Speaker Info**  
   This giving the best results aligns with CosyVoice’s ability to simplify when speaker differentiation isn’t needed. Without prompts or embeddings, the model treats all audio as coming from one speaker, focusing purely on semantic and prosodic quality. This reduces complexity and errors, especially if your dataset is small or lacks diverse speaker data. CosyVoice 2.0’s use of Finite Scalar Quantization (FSQ) and a pretrained Qwen2.5-0.5B LM (arXiv:2412.10117) helps it nail naturalness in simpler scenarios, which explains your success here.


### CosyVoice’s Speaker Embedding Design
To clarify your question: **Yes, CosyVoice 2 and 3 rely on speaker embeddings to capture timbre and style, but only in the flow-matching model, not the text-speech LM.** Here’s how this impacts your setup:
- **CosyVoice 1.0**: Speaker embeddings (x-vectors) were used in both the LM and flow-matching model, capturing timbre, intonation, and style for zero-shot and cross-lingual tasks.
- **CosyVoice 2.0**: The LM drops speaker embeddings to focus on semantic understanding (using Qwen2.5-0.5B), while the flow-matching model uses embeddings to handle timbre and acoustic details. This separation boosts prosody naturalness and cross-lingual performance but makes embeddings critical for speaker-specific output.
- **CosyVoice 3.0**: Builds on 2.0, with an improved speech tokenizer trained on multi-task objectives (e.g., ASR, emotion recognition). Speaker embeddings in the flow-matching model are key for fine-grained control over timbre, style, and even emotion.

With `use_spk_embedding: False`, you’re bypassing this core mechanism, which likely weakens experiment 1 and 2, as the model can’t fully leverage speaker-specific info in the flow-matching stage.

### My Configs for Comparison
Here’s a config I’ve used for multi-speaker SFT with CosyVoice 2.0 I’ve also included contrastive search for inference to boost diversity, since you asked about it earlier. These are based on `libritts/cosyvoice2/run.sh` but tweaked for robustness.

#### Config (`config.yaml`)
```yaml
# Model settings
model:
  use_spk_embedding: True  # Enable for speaker-specific timbre/style
  spk_embedding_dim: 192   # Matches CosyVoice’s x-vector default
  prompt_prefix: "Speaker_{spk_id}<|endofprompt|>"  # For multi-speaker tasks
  prompt_loss_weight: 0.5   # Custom tweak to reduce prompt overfitting
  add_prompt_prob: 0.7      # 70% chance to include prompt in training

# Data settings
data:
  train_data: "path/to/libritts/train-clean-100"  # Your dataset path
  spk2info: "path/to/spk2info.json"  # Speaker metadata (e.g., ID, accent)
  utt2spk: "path/to/utt2spk"         # Speaker mappings
  max_length: 200                    # Max token length
  fsq_levels: [8, 8, 8, 8]          # Finite Scalar Quantization (CosyVoice 2.0)

# Training settings
training:
  batch_size: 16
  learning_rate: 1e-4
  num_epochs: 10
  optimizer: "adamw"
  gradient_accumulation_steps: 4

# Inference settings
inference:
  use_prompt: True                 # Use prompt for multi-speaker inference
  contrastive_search: True        # Enable contrastive search
  contrastive_alpha: 0.6          # Balance probability vs. similarity penalty
  contrastive_top_k: 4            # Top-k candidates
  streaming: True                 # Enable 150ms low-latency streaming
```

#### Key Differences from Your Setup（Some of my "little innovations"）
- **`use_spk_embedding: True`**: Critical for multi-speaker tasks. Enables x-vector embeddings in the flow-matching model to capture timbre, which should improve experiment 1 and 2.
- **`prompt_loss_weight: 0.5`**: Your custom setting is great—lowering the prompt’s influence prevents overfitting, as seen in your experiment 1’s errors.
- **`add_prompt_prob: 0.7`**: Your idea of mixing prompted and non-prompted data is spot-on. It helps the model handle both formats, reducing errors when prompts are absent during inference.
- **`fsq_levels`**: Added to leverage CosyVoice 2.0’s Finite Scalar Quantization, which improves codebook utilization and synthesis quality.
- **Contrastive Search**: Included for inference to reduce repetitive outputs, with `contrastive_alpha: 0.6` and `top_k: 4` as a starting point (inspired by our earlier discussion).


### Suggestions for Your Setup
Based on your experiments and CosyVoice’s design:
1. **Enable Speaker Embeddings**: Set `use_spk_embedding: True` and ensure your `utt2spk` and `spk2info.json` are set up correctly. For experiment 1, this should reduce errors by giving the flow-matching model timbre info to work with, complementing the prompt.
2. **Keep Your Custom Tweaks**: Your `prompt_loss_weight: 0.5` and `add_prompt_prob: 0.7` are smart additions. They make the model less sensitive to prompt format, which could explain why experiment 3 worked well. Try these in experiment 1 with embeddings enabled.
3. **Check Dataset Diversity**: If experiment 2 (with `utt2spk`) underperformed, your dataset might lack enough speakers or audio per speaker. Aim for 2-5 hours per speaker and at least 5-10 speakers (like LibriTTS). If it’s smaller, stick with the single-speaker approach for now.
4. **Test Contrastive Search**: Since you asked about it before, try enabling contrastive search in inference to boost diversity and reduce errors. Use the code snippet above with `penalty_alpha=0.6` and `top_k=4`.
5. **Evaluate Metrics**: Run CosyVoice’s eval scripts to compare WER, Speaker Similarity (SS), and NMOS across your experiments. For example, my multi-speaker setup with embeddings got ~3% WER and 0.85 SS, while single-speaker hit ~2% WER.

### Why Your Third Experiment Won
Your single-speaker setup likely excelled because CosyVoice 2.0’s LM (Qwen2.5-0.5B) is optimized for semantic understanding without speaker embeddings, and FSQ improves token quality. With no need to juggle speakers, the model focuses on clean prosody and pronunciation, especially with `use_spk_embedding: False`. But for multi-speaker or zero-shot tasks, enabling embeddings in the flow-matching model is crucial, as CosyVoice 2 and 3 rely on them for timbre and style (arXiv:2412.10117, arXiv:2505.17589).

### Questions for You
To tailor this further:
- What’s your dataset like? (e.g., number of speakers, hours of audio, languages)
- Are you aiming for multi-speaker or zero-shot tasks, or sticking with single-speaker for now?
- Any specific errors in experiment 1 (e.g., mispronunciations, unnatural prosody)?
- Want a full `run.sh` script or `spk2info.json` example?

Happy tweaking! 🎙️

**Comment by haziyevv** [comment]
@ScottishFold007  Thanks for the detailed breakdown — it’s been extremely helpful.

Here’s some more context and my answers to your earlier questions:
1. Speakers / data: I have 4 speakers — two have ~9 hours each, one has ~5 hours, and one has ~1.5 hours.
2. Goal: I’m aiming for a multi-speaker setting, but since single-speaker training gave me the best results so far, I’ve been cloning each voice separately on top of that single-speaker base model.
3. Issues observed: I still get frequent mispronunciations and unnatural prosody. Also, I’ve added some non-speech tags (e.g., [gasp]) that aren’t in the tokenizer — they don’t seem to be learned well, and sometimes get misclassified (e.g., [gasp] produces laughter).

Regarding use_spk_embedding: In my cosyvoice2.yaml, it’s under the padding step as:
- use_spk_embedding:  This setting is available in the cosyvoice2.yaml like this:
```
padding:!name:cosyvoice.dataset.processor.padding
    use_spk_embedding: False # change to True during sft
```

I understand this is useful for the flow-matching model, but in my case I’m only training the LLM, not the flow-matching stage.

On the config suggestions you mentioned:
1.	prompt_loss_weight and add_prompt_prob — I don’t see these in the default configs. Are you adding them directly in the code?
2.	fsq_levels — I also don’t see this anywhere; where is this set?
3.	The inference parameters you shared (use_prompt, contrastive_search, contrastive_alpha, contrastive_top_k, streaming) — I haven’t seen these in the configs either. Where should I tweak these?

Any pointers on where to hook these in would be appreciated. Thanks again for sharing your insights — it’s been very valuable.

**Comment by ScottishFold007** [comment]
> [@ScottishFold007](https://github.com/ScottishFold007) Thanks for the detailed breakdown — it’s been extremely helpful.
> 
> Here’s some more context and my answers to your earlier questions:
> 
> 1. Speakers / data: I have 4 speakers — two have ~9 hours each, one has ~5 hours, and one has ~1.5 hours.
> 2. Goal: I’m aiming for a multi-speaker setting, but since single-speaker training gave me the best results so far, I’ve been cloning each voice separately on top of that single-speaker base model.
> 3. Issues observed: I still get frequent mispronunciations and unnatural prosody. Also, I’ve added some non-speech tags (e.g., [gasp]) that aren’t in the tokenizer — they don’t seem to be learned well, and sometimes get misclassified (e.g., [gasp] produces laughter).
> 
> Regarding use_spk_embedding: In my cosyvoice2.yaml, it’s under the padding step as:
> 
> * use_spk_embedding:  This setting is available in the cosyvoice2.yaml like this:
> 
> ```
> padding:!name:cosyvoice.dataset.processor.padding
>     use_spk_embedding: False # change to True during sft
> ```
> 
> I understand this is useful for the flow-matching model, but in my case I’m only training the LLM, not the flow-matching stage.
> 
> On the config suggestions you mentioned:
> 
> 1. prompt_loss_weight and add_prompt_prob — I don’t see these in the default configs. Are you adding them directly in the code?
> 2. fsq_levels — I also don’t see this anywhere; where is this set?
> 3. The inference parameters you shared (use_prompt, contrastive_search, contrastive_alpha, contrastive_top_k, streaming) — I haven’t seen these in the configs either. Where should I tweak these?
> 
> Any pointers on where to hook these in would be appreciated. Thanks again for sharing your insights — it’s been very valuable.

These are all my own "hacked" versions built upon the official repo, consisting of new features or components. For example, I introduced contrastive_search during inference, which significantly improves the emotional richness and prosody naturalness of the generated audio. Here's an example:

```python
# Contrastive version
from contrastive_search_tts import ContrastiveSearchTTS
from cosyvoice.utils.file_utils import load_wav

audio_path = ".voice_female.mp3"
prompt_speech_16k = load_wav(audio_path, 16000)
tts_text = "xxx"

# 1. Create wrapper (automatically detects your model architecture)
cs_wrapper = ContrastiveSearchTTS(model)

# 2. Uses exactly the same interface as original, just with added Contrastive Search params
output = cs_wrapper.inference_zero_shot(
    tts_text,
    prompt_speech_16k,
    penalty_alpha=0.9,    # Repetition penalty strength (0.0-1.0)
    top_k=38,             # Number of candidate tokens
    mode="adaptive"       # "standard" or "adaptive"
)
display(Audio(output['tts_speech'], rate=22050))
```
Note:You can refer to the implementation here: [Contrastive Search Is What You Need For Neural Text Generation](https://github.com/yxuansu/Contrastive_Search_Is_What_You_Need)

**Comment by haziyevv** [comment]
@ScottishFold007  Seems like a great effort, thanks a lot

**Comment by alexlfyang** [comment]
>   prompt_prefix: "Speaker_{spk_id}<|endofprompt|>"  # For multi-speaker tasks
>   prompt_loss_weight: 0.5   # Custom tweak to reduce prompt overfitting

@ScottishFold007 @haziyevv May I ask what code you employed to make this experiments for multi-speaker fine-tuning. It seems there is not any option about prompt in the config file of current repo.

**Comment by haziyevv** [comment]
For my part I can answer like this. During finetuning and inference, I am adding speaker name to the prompt. 
For example, I have 3 speakers Anna, Mike and Maria. I create the samples like this:
You are Anna speak with joy<|endofprompt|>Hello, how are you?
You are Mike speak angry<|endofprompt|>Hey what's up?
etc.,

During inference, for each speaker I add `You are speaker_name` infront of the promt. Making this addition during finetuning teaches the LLM to differentiate between the speakers and you get different prosodies for each speaker.

**Comment by ferb2015** [comment]
The training with spkid prompt provides too weak a constraint, and using SFT during inference suffers from cold-start issues, resulting in higher WER.
